atom/browser/ui/win/thumbar_host.cc

```diff
@@ -0,0 +1,139 @@
+const int kButtonIdBase = 40001;
+
+bool GetThumbarButtonFlags(const std::vector<std::string>& flags,
+                           THUMBBUTTONFLAGS* out) {
+  if (flags.empty()) {
+    *out = THBF_ENABLED;
+    return true;
+  }
+  THUMBBUTTONFLAGS result = static_cast<THUMBBUTTONFLAGS>(0);
+  for (const auto& flag : flags) {
+    if (flag == "enabled") {
+      result |= THBF_ENABLED;
+    } else if (flag == "disabled") {
+      result |= THBF_DISABLED;
+    } else if (flag == "dismissonclick") {
+      result |= THBF_DISMISSONCLICK;
+    } else if (flag == "nobackground") {
+      result |= THBF_NOBACKGROUND;
+    } else if (flag == "hidden") {
+      result |= THBF_HIDDEN;
+    } else if (flag == "noninteractive") {
+      result |= THBF_NONINTERACTIVE;
+    } else {
+      return false;
+    }
+  }
+  *out = result;
+  return true;
+}
+
+}  // namespace
+
+ThumbarHost::ThumbarHost(HWND window) : is_initialized_(false),
+                                        window_(window) {
+}
+
+ThumbarHost::~ThumbarHost() {
+}
+
+bool ThumbarHost::SetThumbarButtons(
+    const std::vector<ThumbarHost::ThumbarButton>& buttons) {
+  if (buttons.size() > kMaxButtonsCount)
+    return false;
+
+  base::win::ScopedComPtr<ITaskbarList3> taskbar;
+  if (FAILED(taskbar.CreateInstance(CLSID_TaskbarList,
+                                    nullptr,
+                                    CLSCTX_INPROC_SERVER)) ||
+      FAILED(taskbar->HrInit())) {
+    return false;
+  }
+
+  THUMBBUTTON thumb_buttons[kMaxButtonsCount] = {};
+  thumbar_button_clicked_callback_map_.clear();
+
+  // Once a toolbar with a set of buttons is added to thumbnail, there is no way
+  // to remove it without re-creating the window.
+  // To achieve to re-set thumbar buttons,  we initialize the buttons with
+  // HIDDEN state and only updated the caller's specified buttons.
+  //
+  // Initialize all thumb buttons with HIDDEN state.
+  for (int i = 0; i < kMaxButtonsCount; ++i) {
+    thumb_buttons[i].iId = kButtonIdBase + i;
+    thumb_buttons[i].dwMask = THB_FLAGS;  // dwFlags is valid.
+    thumb_buttons[i].dwFlags = THBF_HIDDEN;
+  }
+
+  // Update the callers' specified buttons.
+  for (size_t i = 0; i < buttons.size(); ++i) {
+    if (!GetThumbarButtonFlags(buttons[i].flags, &thumb_buttons[i].dwFlags))
+      return false;
+    thumb_buttons[i].dwMask = THB_ICON | THB_FLAGS;
+    thumb_buttons[i].hIcon = IconUtil::CreateHICONFromSkBitmap(
+        buttons[i].icon.AsBitmap());
+    if (!buttons[i].tooltip.empty()) {
+      thumb_buttons[i].dwMask |= THB_TOOLTIP;
+      wcscpy_s(thumb_buttons[i].szTip,
+               base::UTF8ToUTF16(buttons[i].tooltip).c_str());
+    }
+
+    thumbar_button_clicked_callback_map_[thumb_buttons[i].iId] =
+        buttons[i].clicked_callback;
+  }
+
+  bool is_success = false;
+  if (!is_initialized_) {
+    is_initialized_ = true;
+    is_success = taskbar->ThumbBarAddButtons(
+        window_, kMaxButtonsCount, thumb_buttons) == S_OK;
```

Comment:
Done. Change to the reset button size. But we need to be always set `kMaxButtonsCount` in the following update to make a totally clean and reset, considering such a scenario:

1. The user set 7 buttons at first.
2. The user set 2 buttons again. 
3. Then the thumbnail toolbar will still show 7 buttons with the 2 buttons updated.